invenio_rdm_records/services/access/service.py

```diff
@@ -0,0 +1,761 @@
+            expand=expand,
+        )
+
+    def read_all_grants(self, identity, id_, expand=False):
+        """Read the access grants of a record (resp. its parent)."""
+        record, parent = self.get_parent_and_record_or_draft(id_)
+
+        # Permissions
+        self.require_permission(identity, "manage", record=record)
+
+        # Fetching
+        return self.grant_result_list(
+            service=self,
+            identity=identity,
+            results=parent.access.grants,
+            expand=expand,
+        )
+
+    @unit_of_work()
+    def delete_grant(self, identity, id_, grant_id, uow=None):
+        """Delete an access grant for a record (resp. its parent)."""
+        record, parent = self.get_parent_and_record_or_draft(id_)
+
+        # Permissions
+        self.require_permission(identity, "manage", record=record)
+
+        # Fetching
+        if not 0 <= grant_id < len(parent.access.grants):
+            raise LookupError(str(grant_id))
+
+        # Deletion
+        parent.access.grants.pop(grant_id)
+
+        # Commit
+        uow.register(RecordCommitOp(parent))
+        if record:
+            uow.register(RecordCommitOp(record))
+
+        # Index all child records of the parent
+        self._index_related_records(record, parent, uow=uow)
+
+        return True
+
+    #
+    # Access requests
+    #
+
+    @unit_of_work()
+    def create_user_access_request(
+        self, identity, id_, message, expand=False, uow=None
+    ):
+        """Create a user access request for the given record."""
+        # TODO hide behind "authenticated_user" permission wall
+        # TODO deny if the identity already has access?
+        # TODO detect duplicate requests
+        record = self.record_cls.pid.resolve(id_)
+        self.require_permission(identity, "read", record=record)  # TODO
+
+        # TODO check that the user cannot actually access files
+        #      -> probably better with a new separate permission policy entry
+        denied = False
+        try:
+            self.require_permission(identity, "read_files", record=record)
+        except PermissionDeniedError:
+            denied = True
+
+        if not denied:
+            raise PermissionDeniedError()
+
+        # TODO refactor the duplicate detection logic
+        req_cls = current_requests_service.record_cls
+        model_cls = req_cls.model_cls
+        requests = [
+            request
+            for request in (
+                req_cls(rm.data, model=rm)
+                for rm in model_cls.query.filter(
+                    model_cls.json["created_by"] == {"user": str(identity.id)},
+                    model_cls.json["topic"] == {"record": id_},
+                )
+                if rm.data and rm.data["type"] == UserAccessRequest.type_id
+            )
+            if request.is_open
+        ]
+
+        if any((req.is_open for req in requests)):
+            raise DuplicateAccessRequestError([str(r.id) for r in requests])
```

Comment:
The DB doesn't have such elaborate checks for duplicates, and given that there's some checking of the request status[x] involved I'm not 100% sure it could even really have them.
Without this check, it was entirely possible to have multiple access requests open in parallel even though that doesn't really make sense here.
Maybe it makes sense for a different type of request (potentially even just a future one), but definitely not here :smile: 

[x] We're checking for `is_open`, which depends on the [request type](https://github.com/inveniosoftware/invenio-requests/blob/master/invenio_requests/customizations/request_types.py#L54-L62) in question.

> Might you can just check if requests is empty, given that you are already filtering above request.is_open?

Good catch, this was a leftover :smile: